File: src/custos/eval/harness/assistants/user_confirmation.py
"""A2 ``user_confirmation`` — prompts the user for every denied call.

No LLM. Max security / severe fatigue. Clean-room re-implementation of the
documented behaviour of ``Janus/src/permissions/assistants/user_confirmation.py``:
log the call, ask yes/no, return ``approve_once`` on yes else ``reject``.
Does NOT persist rules.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from custos.eval.harness.assistants.base import BasePermissionAssistant
from custos.eval.harness.schema import JanusAssistantOutput, JanusAssistantVerdict

__all__ = ["UserConfirmationAssistant"]


class UserConfirmationAssistant(BasePermissionAssistant):
    name = "user_confirmation"

    async def handle_permission_denial(
        self,
        subject: dict[str, Any],
        tool_name: str,
        action: str,
        args: dict[str, Any],
        failed_policies: Sequence[Mapping[str, Any]],
    ) -> JanusAssistantOutput:
        prompt = f"Allow tool call {tool_name}.{action}?\n  args: {args}"
        approved = await self._confirm(prompt)
        if approved:
            return JanusAssistantOutput(
                decision=JanusAssistantVerdict.APPROVE_ONCE,
                reason="user approved one-time tool call",
            )
        return JanusAssistantOutput(
            decision=JanusAssistantVerdict.REJECT,
            reason="user rejected tool call",
        )

File: packages/custos-ts/src/assistants/user_confirmation.ts
// A2 — `user-confirmation`. Prompts the user for every denied call.
// Max security, severe fatigue. No LLM. Baseline.

// In Custos semantics this assistant emits `prompt` (a Custos-only
// extension — Janus A2 maps to it via the contract's verdict mapping
// for the "escalate" leg, but Janus's `approve_once`/`reject` doesn't
// have a deterministic "always ask" output so we synthesize `prompt`).

import type { Assistant } from "./base.ts";
import type { AssistantOutput, Invocation, SubjectContext } from "../schema.ts";

export class UserConfirmationAssistant implements Assistant {
  readonly name = "user-confirmation";
  readonly exfiltratesArgs = false;

  decide(_inv: Invocation, _ctx: SubjectContext): AssistantOutput {
    return {
      decision: "prompt",
      risk: 0.5,
      reasoning: "A2 user-confirmation: always ask the user",
      fatigue_hint: false,
      persist_rule: null,
    };
  }
}